src/packets/mod.rs

```diff
@@ -2762,6 +2763,171 @@ impl MySerialize for ComStmtClose {
+}
+
+impl ComStmtBulkExecuteRequestBuilder {
+    pub fn new(stmt_id: u32, max_payload: usize) -> Self {
+        Self {
+            stmt_id,
+            with_types: true,
+            paramset: Vec::new(),
+            payload_len: 0,
+            max_payload_len: max_payload,
+        }
+    }
+
+    pub fn next(&mut self) -> () {
+        self.with_types = false;
+        self.paramset.clear();
+        self.payload_len = 0;
+    }
+    pub fn add_row(&mut self, params: &[Value]) -> bool {
+        if self.with_types && self.payload_len == 0 {
+            self.payload_len = params.len() * 2;
+        }
+        let mut data_len = 0;
+        for p in params {
+            match p.bin_len() as usize {
+                0 => data_len += 1,     // NULLs take 1 byte for the indicator
+                x => data_len += x + 1, // non-NULLs take their length + 1 byte for the indicator
+            }
+        }
+        // It should be really total packet len(+7 + 4)compared against max allowed packet size, not MAX_PAYLOAD_LEN
+        if 7 + self.payload_len + data_len > self.max_payload_len {
+            return true;
+        }
+        self.paramset.push(params.to_vec());
+        self.payload_len += data_len;
+        false
+    }
+
+    pub fn has_rows(&self) -> bool {
+        !self.paramset.is_empty()
+    }
+
+    pub fn build(&self) -> ComStmtBulkExecuteRequest<'_> {
+        ComStmtBulkExecuteRequest {
+            com_stmt_bulk_execute: ConstU8::new(),
+            stmt_id: RawInt::new(self.stmt_id),
+            bulk_flags: if self.with_types {
+                Const::new(StmtBulkExecuteParamsFlags::SEND_TYPES_TO_SERVER)
+            } else {
+                Const::new(StmtBulkExecuteParamsFlags::empty())
+            },
+            params: &self.paramset,
+        }
+    }
+}
+
+define_header!(
+    ComStmtBulkExecuteHeader,
+    COM_STMT_BULK_EXECUTE,
+    InvalidComStmtBulkExecuteHeader
+);
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ComStmtBulkExecuteRequest<'a> {
+    com_stmt_bulk_execute: ComStmtBulkExecuteHeader,
+    stmt_id: RawInt<LeU32>,
+    bulk_flags: Const<StmtBulkExecuteParamsFlags, LeU16>,
+    // max params / bits per byte = 8192
+    params: &'a Vec<Vec<Value>>,
+}
+
+impl<'a> ComStmtBulkExecuteRequest<'a> {
+    pub fn stmt_id(&self) -> u32 {
+        self.stmt_id.0
+    }
+
+    pub fn bulk_flags(&self) -> StmtBulkExecuteParamsFlags {
+        self.bulk_flags.0
+    }
+
+    pub fn params(&self) -> &[Vec<Value>] {
+        self.params.as_ref()
+    }
+}
+
+impl MySerialize for ComStmtBulkExecuteRequest<'_> {
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        self.com_stmt_bulk_execute.serialize(&mut *buf);
+        self.stmt_id.serialize(&mut *buf);
+        self.bulk_flags.serialize(&mut *buf);
+
+        if self
+            .bulk_flags
+            .0
+            .contains(StmtBulkExecuteParamsFlags::SEND_TYPES_TO_SERVER)
+        {
+            for param in &self.params[0] {
```

Comment:
Hmm. This will panic if `SEND_TYPES_TO_SERVER` is set and the `params` is empty that seems reachable from `ComStmtBulkExecuteRequestBuilder` 🤔